In [3]:
#importa las librerias necesarias
import pandas as pd 
import numpy as np 
import seaborn as sns
import matplotlib.pyplot as plt

from src import soporte_limpieza as sl

In [4]:
#configura para poder visualizar todas las columnas del dataframe
pd.set_option("display.max_columns", None)

In [5]:
#lee el pickle que contiene el df final 
df = pd.read_pickle("datos/datos-final.pkl")
df.head()

CÓDIGO ÓRGÃO SUPERIOR       NOME ÓRGÃO SUPERIOR  CÓDIGO ÓRGÃO  \
0                  63000  Advocacia-Geral da União         63000   
1                  63000  Advocacia-Geral da União         63000   
2                  63000  Advocacia-Geral da União         63000   
3                  63000  Advocacia-Geral da União         63000   
4                  63000  Advocacia-Geral da União         63000   

                                          NOME ÓRGÃO  CÓDIGO UNIDADE GESTORA  \
0  Advocacia-Geral da União - Unidades com víncul...                  110060   
1  Advocacia-Geral da União - Unidades com víncul...                  110060   
2  Advocacia-Geral da União - Unidades com víncul...                  110060   
3  Advocacia-Geral da União - Unidades com víncul...                  110060   
4  Advocacia-Geral da União - Unidades com víncul...                  110060   

                            NOME UNIDADE GESTORA CATEGORIA ECONÔMICA  \
0  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   
1  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   
2  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   
3  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   
4  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   

              ORIGEM RECEITA                                ESPÉCIE RECEITA  \
0  Outras Receitas Correntes  Bens, Direitos e Valores Incorporados ao Patr   
1  Outras Receitas Correntes    Indenizações, restituições e ressarcimentos   
2  Outras Receitas Correntes  Multas administrativas, contratuais e judicia   
3  Outras Receitas Correntes  Bens, Direitos e Valores Incorporados ao Patr   
4  Outras Receitas Correntes    Indenizações, restituições e ressarcimentos   

                                    DETALHAMENTO  VALOR PREVISTO ATUALIZADO  \
0  REC.DIVIDA ATIVA NAO TRIBUTARIA DE OUTRAS REC                        0.0   
1   RECUPERACAO DE DESPESAS DE EXERC. ANTERIORES                        0.0   
2                  OUTRAS MULTAS E JUROS DE MORA                        0.0   
3      REC.DIV.ATIVA POR INFRAÇÃO ADMINISTRATIVA                        0.0   
4                            OUTRAS RESTITUICOES                        0.0   

   VALOR LANÇADO  VALOR REALIZADO  PERCENTUAL REALIZADO DATA LANÇAMENTO  \
0            0.0          1297.13                   0.0      2013-12-31   
1            0.0      26666621.42                   0.0      2013-12-31   
2            0.0        301251.13                   0.0      2013-12-31   
3            0.0          1855.58                   0.0      2013-12-31   
4            0.0         52140.68                   0.0      2013-12-31   

   ANO EXERCÍCIO  
0           2013  
1           2013  
2           2013  
3           2013  
4           2013

Antes de nada, voy a normalizar las columnas. En primer lugar, los nombres de las columnas voy a dejarlos todos por igual: en mayuscula la primera letra de la primera palabra y el resto todo minúsculas. Además, en vez de espacios entre las palabras, voy a poner guión bajo

In [6]:
#saca el nombre de las columnas
df.columns

Index(['CÓDIGO ÓRGÃO SUPERIOR', 'NOME ÓRGÃO SUPERIOR', 'CÓDIGO ÓRGÃO',
       'NOME ÓRGÃO', 'CÓDIGO UNIDADE GESTORA', 'NOME UNIDADE GESTORA',
       'CATEGORIA ECONÔMICA', 'ORIGEM RECEITA', 'ESPÉCIE RECEITA',
       'DETALHAMENTO', 'VALOR PREVISTO ATUALIZADO', 'VALOR LANÇADO',
       'VALOR REALIZADO', 'PERCENTUAL REALIZADO', 'DATA LANÇAMENTO',
       'ANO EXERCÍCIO'],
      dtype='object')

In [7]:
#diccionario pasado por chatgpt con las indicaciones mias en el renombrado de las columnas 
renombrado_columnas = {
    'CÓDIGO ÓRGÃO SUPERIOR': 'Código_órgão_superior',
    'NOME ÓRGÃO SUPERIOR': 'Nome_órgão_superior',
    'CÓDIGO ÓRGÃO': 'Código_órgão',
    'NOME ÓRGÃO': 'Nome_órgão',
    'CÓDIGO UNIDADE GESTORA': 'Código_unidade_gestora',
    'NOME UNIDADE GESTORA': 'Nome_unidade_gestora',
    'CATEGORIA ECONÔMICA': 'Categoria_econômica',
    'ORIGEM RECEITA': 'Origem_receita',
    'ESPÉCIE RECEITA': 'Espécie_receita',
    'DETALHAMENTO': 'Detalhamento',
    'VALOR PREVISTO ATUALIZADO': 'Valor_previsto_atualizado',
    'VALOR LANÇADO': 'Valor_lançado',
    'VALOR REALIZADO': 'Valor_realizado',
    'PERCENTUAL REALIZADO': 'Percentual_realizado',
    'DATA LANÇAMENTO': 'Data_lançamento',
    'ANO EXERCÍCIO': 'Ano_exercício'
}


In [8]:
#renombrado de columnas
df.rename(columns=renombrado_columnas, inplace=True)
df.sample()

Código_órgão_superior     Nome_órgão_superior  Código_órgão  \
851556                  26000  Ministério da Educação         26280   

                                         Nome_órgão  Código_unidade_gestora  \
851556  Fundação Universidade Federal de São Carlos                  154049   

                               Nome_unidade_gestora Categoria_econômica  \
851556  FUNDACAO UNIVERSIDADE FEDERAL DE SAO CARLOS  Receitas Correntes   

             Origem_receita                               Espécie_receita  \
851556  Receita de Serviços  Serviços Administrativos e Comerciais Gerais   

                                       Detalhamento  \
851556  SERV.ADMINISTRAT.E COMERCIAIS GERAIS-PRINC.   

        Valor_previsto_atualizado  Valor_lançado  Valor_realizado  \
851556                        0.0            0.0            840.0   

        Percentual_realizado Data_lançamento  Ano_exercício  
851556                   0.0      2020-10-06           2020

Ahora, por otro lado, a simple vista se observa que hay algunas columnas cuyo contenido está en mayúsculas entero, otras combinan mayúsculas con minúsculas. Para que toda la base de datos este más homogénea visualmente he decidido que voy a generar una función que vaya iterando por cada columna y deje el contenido de la siguiente forma: primera palabra en mayúscula la primera letra, el resto de las palabras todas en minúscula. Además, voy a eliminar espacios al principio y al final de cada string para asegurar que no hay ningún espacio donde no tenga que estar

In [9]:
#se genera la funcion que va a capitalizar el contenido de cada registro por cada columna
def normalizar_columnas(columna):
    return columna.str.capitalize().str.strip()

In [10]:
#se aplica la funcion a las columnas de tipo string
df[["Nome_órgão_superior", "Nome_órgão", "Nome_unidade_gestora", "Categoria_econômica", "Origem_receita", "Espécie_receita", "Detalhamento"]] = df[["Nome_órgão_superior", "Nome_órgão", "Nome_unidade_gestora", "Categoria_econômica", "Origem_receita", "Espécie_receita", "Detalhamento"]].apply(normalizar_columnas)
df.sample()

Código_órgão_superior                     Nome_órgão_superior  \
882034                  53000  Ministério do desenvolvimento regional   

        Código_órgão                                       Nome_órgão  \
882034         53202  Superintendência do desenvolvimento da amazônia   

        Código_unidade_gestora                         Nome_unidade_gestora  \
882034                  533013  Superintend. do desenvolvimento da amazonia   

       Categoria_econômica       Origem_receita      Espécie_receita  \
882034  Receitas correntes  Receita patrimonial  Valores mobiliários   

                                        Detalhamento  \
882034  Remuneracao de depositos bancarios-principal   

        Valor_previsto_atualizado  Valor_lançado  Valor_realizado  \
882034                        0.0            0.0         80935.68   

        Percentual_realizado Data_lançamento  Ano_exercício  
882034                   0.0      2020-01-09           2020

Ahora ya, a pesar de que siga en portugués, visualmente parece otra cosa

In [11]:
#funcion para obtener un reporte que de la informacion del numero de nulos, del porcentaje de nulos y del tipo de dato (por cada columna)
def info_reporte(dataframe):
    df_report = pd.DataFrame()
    df_report["Numero_nulos"] = dataframe.isnull().sum()
    df_report["Porcentaje_nulos"] = round((dataframe.isnull().sum()/dataframe.shape[0]*100), 2)
    df_report["Tipo_dato"] = dataframe.dtypes
    return df_report

In [12]:
#invoco a la funcion de reporte
info_reporte(df)

Numero_nulos  Porcentaje_nulos       Tipo_dato
Código_órgão_superior                 0              0.00           int64
Nome_órgão_superior                   0              0.00          object
Código_órgão                          0              0.00           int64
Nome_órgão                            0              0.00          object
Código_unidade_gestora                0              0.00           int64
Nome_unidade_gestora                  0              0.00          object
Categoria_econômica                   0              0.00          object
Origem_receita                        0              0.00          object
Espécie_receita                       0              0.00          object
Detalhamento                          0              0.00          object
Valor_previsto_atualizado             0              0.00         float64
Valor_lançado                         0              0.00         float64
Valor_realizado                       0              0.00         float64
Percentual_realizado                  0              0.00         float64
Data_lançamento                     578              0.06  datetime64[ns]
Ano_exercício                         0              0.00           int64

In [13]:
#comprueba si hay o no duplicados
df.duplicated().sum()

np.int64(0)

Como era de esperar por el análsis exploratorio individual ya hecho, el tipo de dato de las columnas ya es correcto, pues ya se corrigió en cada dataframe individual por año, no hay duplicados y solo hay nulos en la columna de fecha de registro, pero es un porcentaje mínimo (0,06%) y se van a mantener como están

Antes de continuar con el análisis exploratorio, hay 3 columnas las cuales creo que dan información redundante y que no son necesarias en el análisis. Las columnas Código_órgão_superior, Código_órgão y Código_unidade_gestora son 3 columnas de las que se puede prescindir, pues disponemos de la descripción de cada código en otras 3 columnas más, y son estas 3 las que se van a emplear para los análisis, pues las columnas de códigos no se interpretan bien, ni son apropiadas para posteriormente la visualización y explicación de resultados. En caso de estar trabajando con consultas de SQL estas 3 columnas podrían ser interesantes en caso de tener que realizar algún join (al fin y al cabo es más sencillo unir por códigos que por un texto específico en el que hay más probabilidades de equivocación a la hora de escribirlo), pero como no es el caso, voy a pasar a eliminarlas y a mantener las 3 (Nome_órgão_superior, Nome_órgão, Nome_unidade_gestora) que verdaderamente interesan

In [14]:
len(df["Código_órgão_superior"].unique())

25

In [15]:
len(df["Nome_órgão_superior"].unique())

25

In [16]:
len(df["Código_órgão"].unique())

291

In [17]:
len(df["Nome_órgão"].unique())

287

In [18]:
len(df["Código_unidade_gestora"].unique())

364

In [19]:
len(df["Nome_unidade_gestora"].unique())

356

In [20]:
#elimina las columnas de codigos
df.drop(columns=["Código_órgão_superior", "Código_órgão", "Código_unidade_gestora"], inplace=True)

In [21]:
#saca los descriptivos de las variables numéricas
df.describe().T

count                           mean  \
Valor_previsto_atualizado  1026299.0                27611421.049683   
Valor_lançado              1026299.0                 5242806.106963   
Valor_realizado            1026299.0                24980422.309138   
Percentual_realizado       1026299.0                      80.313812   
Data_lançamento              1025721  2018-09-20 12:23:18.569786368   
Ano_exercício              1026299.0                    2018.225555   

                                           min                  25%  \
Valor_previsto_atualizado         -214773743.0                  0.0   
Valor_lançado             -148347267246.910004                  0.0   
Valor_realizado           -156285934188.029999              260.905   
Percentual_realizado                 -114552.0                  0.0   
Data_lançamento            2013-12-31 00:00:00  2017-03-31 00:00:00   
Ano_exercício                           2013.0               2017.0   

                                           50%                  75%  \
Valor_previsto_atualizado                  0.0                  0.0   
Valor_lançado                              0.0                  0.0   
Valor_realizado                        3022.32            40853.855   
Percentual_realizado                       0.0                  0.0   
Data_lançamento            2018-08-27 00:00:00  2020-02-12 00:00:00   
Ano_exercício                           2018.0               2020.0   

                                           max                std  
Valor_previsto_atualizado      1603521711208.0  3393367175.127452  
Valor_lançado              357160677863.969971   810883695.355811  
Valor_realizado            771117711060.949951    1406892548.8405  
Percentual_realizado                72363772.0       71481.112902  
Data_lançamento            2021-12-03 00:00:00                NaN  
Ano_exercício                           2021.0           1.737966

Lo primero que me llama la atención, que ya es algo que había observado, son los ceros de las columnas Valor_previsto_atualizado, Valor_lançado, Percentual_realizado

Valor_previsto_atualizado

In [22]:
#registros en los que el el valor previsto actualizado es distinto de cero
df[df["Valor_previsto_atualizado"] != 0].shape[0] / df.shape[0] *100

1.93315982964029

Tan solo hay un 1,93% de los registros donde el valor previsto actualizado es distinto de cero 

Es bastante curioso que a pesar de que prácticamente en el 98% de los registros el valor previsto actualizado es cero, la media tome un valor grande, como lo es: 27611421 y la mediana un valor de cero. Esto parece indicar que hay presencia de outliers en aquellos registros contados sobre los que disponemos de información en el valor previsto actualizado y que la media está sesgada y no es un valor representativo. El mínimo es de -214773743, un valor previsto muy bajo. Aún más llama la atención el máximo, que es de 1603521711208, un valor muy elevado. Además, la desviación estándar es muy alta, lo que es indicador de elevada dispersión en los datos

Por lo que parece, con la columna de Valor_lançado y Percentual_realizado ocurre parecido. El número de registros con un valor distinto de cero en estas columnas debe ser mínimo, y además, por el mínimo y el máximo, que son valores o extremadamente bajos, o extremadamente altos, parece que hay unos outliers que distorsionan completamente la media

Valor_lançado

In [23]:
#porcentaje de registros cuyo Valor_lançado es distinto de cero 
df[df["Valor_lançado"] != 0].shape[0] / df.shape[0] *100

0.7033038130213515

Tan solo el 0,7% de los registros tiene un valor lanzado distinto de cero 

Tiene una media de 5242806 y una mediana de 0. La media está completamente sesgada por valores extremos. Hay un mínimo de -148347267246 y un máximo de 357160677863. Por el valor de la desviación estándar, parece que hay elevada dispersión en los datos

Percentual_realizado

In [24]:
#porcentaje de registros cuyo Percentual_realizado es distinto de cero 
df[df["Percentual_realizado"] != 0].shape[0] / df.shape[0] *100

0.7299042481771881

Tan solo el 0,72% de los registros tienen un valor de percentual realizado distinto de cero

Tiene una media de 80% y una mediana de 0. La media está completamente sesgada por valores extremos. Hay un mínimo de -114552% y un máximo de 72363772%. Por el valor de la desviación estándar, parece que hay elevada dispersión en los datos

## Me queda por hablar de la de valor realizado y de la de fehca de registro y de las variables categoricas

## Me queda aquí hacer análisis super basico de ingresos (total por años y en total de todos los años). tambien por ejemplo por categoria economica, por nombre de organo superiror

Observación y entendimiento de la cantidad de ceros

In [25]:
#devuelve los registros donde Valor_previsto_atualizado es distinto de cero y Ano_exercício es 2016
valor_previsto_no_cero_2016 = df[(df["Valor_previsto_atualizado"] != 0) & (df["Ano_exercício"] == 2016)]
valor_previsto_no_cero_2016

Nome_órgão_superior  \
14095                        Advocacia-geral da união   
15635                    Controladoria-geral da união   
15736   Ministério da agricultura, pecuária e abastec   
15743   Ministério da agricultura, pecuária e abastec   
15846   Ministério da agricultura, pecuária e abastec   
...                                               ...   
207253                       Presidência da república   
207545                       Presidência da república   
207856                       Presidência da república   
208066                       Presidência da república   
208103                       Presidência da república   

                                               Nome_órgão  \
14095   Advocacia-geral da união - unidades com víncul...   
15635                        Controladoria-geral da união   
15736   Ministério da agricultura, pecuária e abasteci...   
15743         Empresa brasileira de pesquisa agropecuária   
15846   Ministério da agricultura, pecuária e abasteci...   
...                                                   ...   
207253                           Presidência da república   
207545                         Fundo de imprensa nacional   
207856                         Fundo de imprensa nacional   
208066     Instituto nacional de tecnologia da informação   
208103              Secretaria da micro e pequena empresa   

                                 Nome_unidade_gestora  \
14095   Coord. geral de orc. fin. e anal. cont. - agu   
15635              Diretoria de gestao interna/se/cgu   
15736       Coord.-geral de orcamento e financas-mapa   
15743                                Embrapa-setorial   
15846       Coord.-geral de orcamento e financas-mapa   
...                                               ...   
207253  Secretaria de administracao-set.orc.e fin./pr   
207545    Fundo de imprensa nacional/exec.orc.financ.   
207856    Fundo de imprensa nacional/exec.orc.financ.   
208066  Instituto nac.de tecnologia da informacao iti   
208103          Secretaria da micro e pequena empresa   

                             Categoria_econômica        Origem_receita  \
14095                         Receitas correntes   Receita de serviços   
15635                        Receitas de capital  Operações de crédito   
15736                         Receitas correntes         Contribuições   
15743                         Receitas correntes   Receita de serviços   
15846                         Receitas correntes  Receita agropecuária   
...                                          ...                   ...   
207253                        Receitas correntes   Receita patrimonial   
207545                        Receitas correntes    Receita industrial   
207856  Receitas correntes - intra-orçamentárias    Receita industrial   
208066                        Receitas correntes   Receita de serviços   
208103                        Receitas correntes   Receita de serviços   

                                     Espécie_receita  \
14095   Serviços administrativos e comerciais gerais   
15635         Operações de crédito - mercado externo   
15736                       Contribuições econômicas   
15743                                Outros serviços   
15846                           Receita agropecuária   
...                                              ...   
207253                           Valores mobiliários   
207545                            Receita industrial   
207856                            Receita industrial   
208066  Serviços administrativos e comerciais gerais   
208103  Serviços administrativos e comerciais gerais   

                                         Detalhamento  \
14095   Inscr.em concursos e proc.seletivos-principal   
15635   Op.de credito contratuais-merc.externo-princ.   
15736   Contr.s/apostas em competicoes hipicas-princ.   
15743                       Outros servicos-principal   
15846                  Receita agropecuaria-principal   
...                        

He decidido observar por ejemplo para el año 2016 aquellos registros donde el valor previsto actualizado es distinto de cero. Tan solo hay 1894 filas. Por lo que observo, parece que lo que hace esta columna es, para cada tipo de ingreso y órgano, recoger una única vez el valor estimado anual que se va a recaudar. Es decir, contamos con el valor previsto para 2016 de cada tipo de ingreso recaudado por cada órgano. Y es por ello por lo que creo que hay tan pocos registros cuyo valor previsto es distinto de cero

Observo que hay registros donde se tiene dato tanto del valor previsto como del valor realizado, y otros donde solo se dispone del valor previsto y el valor realizado es cero

In [26]:
#devuelve las fechas de registro para las que el valor previsto actualizado es distinto de cero
valor_previsto_no_cero_2016["Data_lançamento"].unique()

<DatetimeArray>
['2016-01-21 00:00:00', '2016-06-08 00:00:00', '2016-01-13 00:00:00']
Length: 3, dtype: datetime64[ns]

No hay una única fecha para todos los ingresos en la que se haga esta estimación anual

In [27]:
#filtra por todas las columnas, desde Nome_órgão_superior hasta Detalhamento para que devuelva un ingreso concreto que se ha observado que tiene dato tanto en el valor previsto como en el realizado
df[(df["Nome_órgão_superior"] == "Presidência da república") & (df["Nome_órgão"] == "Fundo de imprensa nacional") & (df["Nome_unidade_gestora"] == "Fundo de imprensa nacional/exec.orc.financ.") & (df["Categoria_econômica"] == "Receitas correntes - intra-orçamentárias") & (df["Origem_receita"] == "Receita industrial") & (df["Espécie_receita"] == "Receita industrial") & (df["Detalhamento"] == "Receita industrial-principal") & (df["Ano_exercício"] == 2016)]

Nome_órgão_superior                  Nome_órgão  \
207607  Presidência da república  Fundo de imprensa nacional   
207608  Presidência da república  Fundo de imprensa nacional   
207609  Presidência da república  Fundo de imprensa nacional   
207610  Presidência da república  Fundo de imprensa nacional   
207611  Presidência da república  Fundo de imprensa nacional   
...                          ...                         ...   
207865  Presidência da república  Fundo de imprensa nacional   
207866  Presidência da república  Fundo de imprensa nacional   
207867  Presidência da república  Fundo de imprensa nacional   
207868  Presidência da república  Fundo de imprensa nacional   
207869  Presidência da república  Fundo de imprensa nacional   

                               Nome_unidade_gestora  \
207607  Fundo de imprensa nacional/exec.orc.financ.   
207608  Fundo de imprensa nacional/exec.orc.financ.   
207609  Fundo de imprensa nacional/exec.orc.financ.   
207610  Fundo de imprensa nacional/exec.orc.financ.   
207611  Fundo de imprensa nacional/exec.orc.financ.   
...                                             ...   
207865  Fundo de imprensa nacional/exec.orc.financ.   
207866  Fundo de imprensa nacional/exec.orc.financ.   
207867  Fundo de imprensa nacional/exec.orc.financ.   
207868  Fundo de imprensa nacional/exec.orc.financ.   
207869  Fundo de imprensa nacional/exec.orc.financ.   

                             Categoria_econômica      Origem_receita  \
207607  Receitas correntes - intra-orçamentárias  Receita industrial   
207608  Receitas correntes - intra-orçamentárias  Receita industrial   
207609  Receitas correntes - intra-orçamentárias  Receita industrial   
207610  Receitas correntes - intra-orçamentárias  Receita industrial   
207611  Receitas correntes - intra-orçamentárias  Receita industrial   
...                                          ...                 ...   
207865  Receitas correntes - intra-orçamentárias  Receita industrial   
207866  Receitas correntes - intra-orçamentárias  Receita industrial   
207867  Receitas correntes - intra-orçamentárias  Receita industrial   
207868  Receitas correntes - intra-orçamentárias  Receita industrial   
207869  Receitas correntes - intra-orçamentárias  Receita industrial   

           Espécie_receita                  Detalhamento  \
207607  Receita industrial  Receita industrial-principal   
207608  Receita industrial  Receita industrial-principal   
207609  Receita industrial  Receita industrial-principal   
207610  Receita industrial  Receita industrial-principal   
207611  Receita industrial  Receita industrial-principal   
...                    ...                           ...   
207865  Receita industrial  Receita industrial-principal   
207866  Receita industrial  Receita industrial-principal   
207867  Receita industrial  Receita industrial-principal   
207868  Receita industrial  Receita industrial-principal   
207869  Receita industrial  Receita industrial-principal   

        Valor_previsto_atualizado  Valor_lançado  Valor_realizado  \
207607                        0.0            0.0        354407.38   
207608                        0.0            0.0        580489.29   
207609                        0.0            0.0        703254.81   
207610                        0.0            0.0        147189.73   
207611                        0.0            0.0        269745.14   
...                           ...            ...              ...   
207865                        0.0            0.0       1157696.18   
207866                        0.0            0.0        388793.40   
207867                        0.0            0.0        875686.39   
207868                        0.0            0.0       1039808.06   
207869                        0.0            0.0        678889.49   

        Percentual_realizado Data_lançamento  Ano_exercício  
207607                   0.0      2016-07-13           2016  
207608                   0.0      2016-08-08

Hay 253 registros de este mismo ingreso que lo recauda el mismo órgano. Un registro contiene la estimación (anual, supongo) y una recaudación. El resto de los registros son recaudaciones de ese ingreso concreto en diferentes fechas. Es decir, hay múltiples recaudaciones reales para este ingreso registradas y tan solo hay una estimación que además coincide en la misma fecha que una de las recaudaciones. En este caso, la fecha en la que aparece la previsión es el 21 de enero. Sin embargo, observamos fechas tanto anteriores (como el 15 de enero) como posteriores (13 de julio por ejemplo) de ejecución de este ingreso. Esto creo que puede ocurrir así porque al fin y al cabo la previsión de cada ingreso está hecha anualmente pero es una previsión actualizada, es decir, es una previsión que puede ser modificada o ajustada a lo largo del año por distintos factores (ajuste por las ejecuciones ya registradas en el año, condiciones económicas como nuevas leyes que hagan modificar la previsión). En los registros donde solo disponemos de la previsión, supongo que en este caso la fecha de registro se refiere más bien a la última fecha donde se ha actualizado esa previsión del ingreso. Si en esta misma fecha hay un valor realizado es porque en la misma fecha donde se actualizó por última vez la previsión del ingreso, se ejecutaría ese ingreso (coincidiría la fecha de previsión con una de las fechas de ejecución del ingreso). Por tanto, es coherente y puede pasar que la fecha de previsión sea posterior, o esté entre medias de las fechas de ejecución de ese ingreso

Por lo tanto, para poder analizar las discrepancias entre lo previsto y lo realmente recaudado, la columna de fecha de registro creo que es bastante irrelevante en el análisis. Si disponemos de una previsión que es anual para cada ingreso, habrá que, anualmente, sacar la recaudación de cada ingreso también, para poder comparar la diferencia entre lo que se esperó que se recaudaría y lo que realmente se recaudó. No tiene sentido comparar la previsión anual con una recaudación mensual, por lo tanto, en este caso, el análisis temporal lo veo coherente hacerlo por años, pero no por meses, y es por ello por lo que considero que la fecha de registro es algo irrelevante

In [28]:
#agrupa por las 7 primeras columnas + año y hace la suma del valor previsto y el valor realizado 
df_estimacion_realizado = df.groupby(["Nome_órgão_superior", "Nome_órgão", "Nome_unidade_gestora", "Categoria_econômica", "Origem_receita", "Espécie_receita", "Detalhamento", "Ano_exercício"])[["Valor_realizado", "Valor_previsto_atualizado"]].sum()
df_estimacion_realizado["Percentual_realizado"] = round((df_estimacion_realizado["Valor_realizado"] / df_estimacion_realizado["Valor_previsto_atualizado"])*100,2).replace(np.inf, "Previsión de 0").replace(-np.inf, "Previsión de 0")
df_estimacion_realizado = df_estimacion_realizado.reset_index()
df_estimacion_realizado

Nome_órgão_superior  \
0      Advocacia-geral da união   
1      Advocacia-geral da união   
2      Advocacia-geral da união   
3      Advocacia-geral da união   
4      Advocacia-geral da união   
...                         ...   
35805  Presidência da república   
35806  Presidência da república   
35807  Presidência da república   
35808  Presidência da república   
35809  Presidência da república   

                                              Nome_órgão  \
0      Advocacia-geral da união - unidades com víncul...   
1      Advocacia-geral da união - unidades com víncul...   
2      Advocacia-geral da união - unidades com víncul...   
3      Advocacia-geral da união - unidades com víncul...   
4      Advocacia-geral da união - unidades com víncul...   
...                                                  ...   
35805  Secretaria de políticas de promoção da igualda...   
35806  Secretaria de políticas de promoção da igualda...   
35807  Secretaria de políticas de promoção da igualda...   
35808  Secretaria de políticas de promoção da igualda...   
35809  Secretaria de políticas de promoção da igualda...   

                                Nome_unidade_gestora Categoria_econômica  \
0      Coord. geral de orc. fin. e anal. cont. - agu  Receitas correntes   
1      Coord. geral de orc. fin. e anal. cont. - agu  Receitas correntes   
2      Coord. geral de orc. fin. e anal. cont. - agu  Receitas correntes   
3      Coord. geral de orc. fin. e anal. cont. - agu  Receitas correntes   
4      Coord. geral de orc. fin. e anal. cont. - agu  Receitas correntes   
...                                              ...                 ...   
35805  Secretaria de polit. de prom.da iguald.racial  Receitas correntes   
35806  Secretaria de polit. de prom.da iguald.racial  Receitas correntes   
35807  Secretaria de polit. de prom.da iguald.racial  Receitas correntes   
35808  Secretaria de polit. de prom.da iguald.racial  Receitas correntes   
35809  Secretaria de polit. de prom.da iguald.racial  Receitas correntes   

                                    Origem_receita  \
0                                    Contribuições   
1      Impostos, taxas e contribuições de melhoria   
2      Impostos, taxas e contribuições de melhoria   
3      Impostos, taxas e contribuições de melhoria   
4                        Outras receitas correntes   
...                                            ...   
35805                    Outras receitas correntes   
35806                    Outras receitas correntes   
35807                    Outras receitas correntes   
35808                    Outras receitas correntes   
35809                    Outras receitas correntes   

                                     Espécie_receita  \
0                           Contribuições econômicas   
1                                              Taxas   
2                                              Taxas   
3                                              Taxas   
4      Bens, direitos e valores incorporados ao patr   
...                                              ...   
35805    Indenizações, restituições e ressarcimentos   
35806    Indenizações, restituições e ressarcimentos   
35807    Indenizações, restituições e ressarcimentos   
35808  Multas administrativas, contratuais e judicia   
35809  Multas administrativas, contratuais e judicia   

                                       Detalhamento  Ano_exercício  \
0             Condecine-divida ativa-multas e juros           2016   
1        Taxa de controle e fisc.ambiental-mul.jur.           2016   
2       Taxa de controle e fisc.ambiental-principal           2016   
3         Taxa de controle e fiscalizacao ambiental           2015   
4         Rec.div.ativa por infracao administrativa           2015   
...                                             ...            ...   
35805                     Restituicoes de convenios           2013   
35806         Restituicoes de convenios - primarias           2014   
35807

De esta forma, para 2016 por ejemplo, lo que estoy logrando es tener para cada ingreso recaudado por cada órgano (que es para lo que tenemos la información anual de la previsión), la previsión anual y la recaudación anual (la suma de todos los meses y días del año). En este punto disponemos de algunos ingresos para los que únicamente dispongo del valor de la previsión. En este caso, será porque se preveía recaudar x cantidad, y finalmente se recaudó cero. También, dispongo de algunos ingresos para los que ocurre el caso contrario, y es que se preveía recaudar 0 pero finalmente se recaudó x cantidad. Por último, dispongo de los que tienen valor distinto de cero tanto en la previsión como en la recaudación real final. En los 3 casos es interesante observar esas discrepancias entre lo que se esperaba recaudar y lo realmente recaudado

En este nuevo dataframe que me he generado con el group by para obtener anualmente la recaudación estimada y real de cada ingreso en cada órgano, voy a añadir dos nuevas columnas, por un lado, el año, y por otro lado, el percentual. Finalmente, cuando obtenga este dataframe nuevo para cada año, los concatenaré, y a partir de ese, ya podré hacer análisis

## hacer analisis por año, por categoria economica, por nombre de organo superior

## si me puedo quedar sin la de data y valor lanzado

In [29]:

"""describe (numericas como objeto)
para las objeto los unicos y value counts
ver que columnas elimino """

'describe (numericas como objeto)\npara las objeto los unicos y value counts\nver que columnas elimino '

In [30]:
"""df_recuento_valores_language = pd.DataFrame(df_netflix_originals["Language"].value_counts().head())
df_recuento_valores_language["%/Total"] = round((df_netflix_originals["Language"].value_counts().head())/(df_netflix_originals.shape[0])*100,2)
df_recuento_valores_language"""

'df_recuento_valores_language = pd.DataFrame(df_netflix_originals["Language"].value_counts().head())\ndf_recuento_valores_language["%/Total"] = round((df_netflix_originals["Language"].value_counts().head())/(df_netflix_originals.shape[0])*100,2)\ndf_recuento_valores_language'

In [31]:
df_estimacion_realizado.columns

Index(['Nome_órgão_superior', 'Nome_órgão', 'Nome_unidade_gestora',
       'Categoria_econômica', 'Origem_receita', 'Espécie_receita',
       'Detalhamento', 'Ano_exercício', 'Valor_realizado',
       'Valor_previsto_atualizado', 'Percentual_realizado'],
      dtype='object')